### DISCUSSION TITLE
What Causes Variable Fixing on the Root Node (Other than Preprocessing/Presolve)

### DISCUSSION THREAD

**Discussion by spkelle2** [coin-or/Cbc#539] [Q&A] [answered]
Hi there,

I'm working with the p0201.mps example and examining the LP ([here's my .lp file](https://github.com/coin-or/Cbc/files/9465769/node0.txt)) corresponding to the first node given to `doOneNode` in the main branch and bound loop (not one of the primal heuristics). I have preprocessing and presolve turned off, so as expected I see the original rows and columns plus rows added from active cutting planes in the LP. What I didn't expect to see was some of the variables already fixed. If preprocessing and presolve are turned off, what else could cause these variable values to become fixed before branching has occurred? Further, are these fixes valid for the convex hull of all feasible solutions or just the optimal solutions?

**Comment by jjhforrest** [comment, marked as answer]
Running that file, heuristics find a solution.  I have not debugged to see exactly what happens, but variables will probably have been fixed to a bound because of their reduced cost.  So the fixes would only be valid for optimal solutions.  If you turn off heuristics, you should see what you expected.

**Reply by spkelle2** [reply]
I apologize for writing that tired at the end of my day. What I meant is that I have heuristics, presolve, and preprocessing all turned off, and I'm getting the .lp generated above from the first node. In that case, would you be able to shed any light on why that's happening?

**Reply by spkelle2** [reply]
(In all fairness and full disclosure) I generated the .lp above in a run where I did forget to turn heuristics off, but I had seen the issue with variables getting fixed while heuristics, presolve, and prepreprocessing were all turned off for another run configuration I had. When I turned heuristics off for the configuration that generated the .lp file I posted above (solving `p0201.mps`), I again got a .lp file for the first node with very similar variable fixing to the above (albeit it appears one or two of the variables are no longer fixed). If it makes a difference, [here's](https://github.com/coin-or/Cbc/files/9472351/node0.txt) the .lp file for the first node with heuristics, presolve, and preprocessing turned off I got when solving `p0201.mps`.

**Reply by spkelle2** [reply]
Just to be overtly clear, here's my run arguments
`-import /Users/sean/coin-or/Data/Sample/p0201.mps -preprocess off -presolve off -heuristics off -log 3 -solve`

Additionally, I'm saving the first node's .lp file right after the event CbcEventHandler::node is called.

**Reply by jjhforrest** [reply]
Round about 3869 of CbcSolver.cpp there is

#ifndef CBC_OTHER_SOLVER
                if (!complicatedInteger && preProcess == 0 && clpSolver->tightenPrimalBounds(0.0, 0, true) != 0) {
                  sprintf(generalPrint, "Problem is infeasible - tightenPrimalBounds!");
So tightenPrimalBounds finds that a variable can not exceed 0.xxx and so fixes it to 0.0

So in that case the fixes are valid for all feasible solutions

**Reply by spkelle2** [reply]
Awesome. That is really great to know, and makes a lot of sense. Thank you for the explanation!
